### PR TITLE
handle inline comments, and ignore spaces outside quotes

### PR DESCRIPTION
**Pull request by lvanderree**
capable of handling inline comments (outside quotes).

So if you want "#" hashes in your values, place them inside quotes.

Also ignoring spaces outside quotes

### CODE REVIEW

**Comment by lvanderree** [comment]
I noticed issues regarding this PR:

- https://github.com/joke2k/django-environ/issues/91
- https://github.com/joke2k/django-environ/pull/475

File: CHANGELOG.rst [hunk 1]
```diff
@@ -15,11 +15,15 @@ Added
   `#463 <https://github.com/joke2k/django-environ/pull/463>`_.
 - Added variable expansion
   `#468 <https://github.com/joke2k/django-environ/pull/468>`_.
+- Added capability to handle comments after #, after quoted values, like `KEY= 'value_1 # value_2' # comment`
```
Comment (sergeyklay) [inline CHANGELOG.rst:18]:
```suggestion
- Added capability to handle comments after #, after quoted values, like ``KEY= 'value_1 # value_2' # comment``
```

File: environ/environ.py [hunk 2]
```diff
@@ -970,9 +970,15 @@ def _keep_escaped_format_characters(match):
             m1 = re.match(r'\A(?:export )?([A-Za-z_0-9]+)=(.*)\Z', line)
             if m1:
                 key, val = m1.group(1), m1.group(2)
-                m2 = re.match(r"\A'(.*)'\Z", val)
+                # search for value within quotes, ignore comments after # (outside quotes)
```
Comment (sergeyklay) [inline environ/environ.py:973]: I paraphrased and reformatted your comment a bit to make it clearer and satisfy the code style.

```suggestion
                # Look for value in quotes, ignore post-# comments
                # (outside quotes)
```

File: environ/environ.py [hunk 3]
```diff
@@ -970,9 +970,15 @@ def _keep_escaped_format_characters(match):
...
+                m2 = re.match(r"\A\s*'(?<!\\)(.*)'\s*(#.*\s*)?\Z", val)
                 if m2:
                     val = m2.group(1)
+                else:
+                    # if no quotes, search for value, ignore comments after first #
```
Comment (sergeyklay) [inline environ/environ.py:978]: I paraphrased and reformatted your comment a bit to make it clearer and satisfy the code style.

```suggestion
                    # For no quotes, find value, ignore comments
                    # after the first #
```

File: tests/test_env.py [hunk 4]
```diff
@@ -175,10 +177,13 @@ def test_mix_tuple_issue_387(self):
...
     def test_str_list_with_spaces(self):
-        assert_type_and_value(list, [' foo', '  bar'],
+        assert_type_and_value(list, [' foo', '  spaces'],
                               self.env('STR_LIST_WITH_SPACES', cast=[str]))
-        assert_type_and_value(list, [' foo', '  bar'],
+        assert_type_and_value(list, [' foo', '  spaces'],
                               self.env.list('STR_LIST_WITH_SPACES'))
+        # assert_type_and_value(list, [' foo', '  quoted'],
+        #                       self.env.list('STR_LIST_WITH_SPACES_QUOTED')
+        #                       ) # TODO: fix this
```
Comment (sergeyklay) [inline tests/test_env.py:186]: Do we really need this piece of code?

**Review by sergeyklay — CHANGES_REQUESTED** [review]
Great work on the PR! Noticed some linter warnings. Please, could you address them? After that, it'll be ready for merge. Thanks!

**Review by sergeyklay — APPROVED** [review]

**Comment by sergeyklay** [comment]
Looks great. Thank you for contributing and patience

**Comment by jonnyhsu** [comment]
> So if you want "#" hashes in your values, place them inside quotes.

Unfortunately the Django `SECRET_KEY` has "#" as a valid character, so this is actually a fairly serious breaking change.

**Comment by lvanderree** [comment]
> Unfortunately the Django `SECRET_KEY` has "#" as a valid character, so this is actually a fairly serious breaking change.

what is the breaking change exactly? Have you got an example that isn't working anymore?
As far as a I see everything is still possible, but with the additional functionality to also be able to place comments

**Comment by sergeyklay** [comment]
.env file contents:
```
SECRET_KEY_499_1=abc#def
SECRET_KEY_499_2='abc#def'
SECRET_KEY_499_3='abc\#def'
SECRET_KEY_499_4="abc\#def"
SECRET_KEY_499_5=abc\#def
```

Test 2750dd54e0e4b1b591a31409cfb1dc16a82b958d (this PR)
```
SECRET_KEY_499_1=abc
SECRET_KEY_499_2=abc#def
SECRET_KEY_499_3=abc\#def
SECRET_KEY_499_4="abc\
SECRET_KEY_499_5=abc\
```

Test 8874288ac11bc70fa491439d52a4c307f9388aea (v0.10.0)
```
SECRET_KEY_499_1=abc#def
SECRET_KEY_499_2=abc#def
SECRET_KEY_499_3=abc\#def
SECRET_KEY_499_4=abc#def
SECRET_KEY_499_5=abc\#def
```

**Comment by sergeyklay** [comment]
Hey @lvanderree,

I'm currently working on some modifications to the functionality you introduced in this PR. While the inline comments feature is useful, it has led to some unintended side effects that need addressing. I've opened a new pull request (see [#500](https://github.com/joke2k/django-environ/pull/500)) to tackle these issues.

I'd appreciate your thoughts and any feedback you might have on this.

**Comment by jonnyhsu** [comment]
> > Unfortunately the Django `SECRET_KEY` has "#" as a valid character, so this is actually a fairly serious breaking change.
> 
> what is the breaking change exactly? Have you got an example that isn't working anymore? As far as a I see everything is still possible, but with the additional functionality to also be able to place comments

If the SECRET_KEY variable starts with a # then the site will not load, because a non-blank secret key is required. If the SECRET_KEY contains a #, then the key will be silently shortened which reduces security.

File: tests/test_env.txt [hunk 5]
```diff
@@ -42,8 +44,11 @@ ESCAPED_VAR=\$baz
 EMPTY_LIST=
 EMPTY_INT_VAR=
 INT_VAR=42
-STR_LIST_WITH_SPACES= foo,  bar
+STR_LIST_WITH_SPACES= foo,  spaces
+STR_LIST_WITH_SPACES_QUOTED=' foo','  quoted'
 STR_VAR=bar
+STR_QUOTED_IGNORE_COMMENT= 'foo' # comment
+STR_QUOTED_INCLUDE_HASH='foo # with hash' # not comment
```
[the remark this answers is not in the document]
Comment (sergeyklay) [reply tests/test_env.txt:51]: See https://github.com/joke2k/django-environ/pull/500

**Pull request merged**